[iOS] Displayed values for date controls may cause overflow onto multiple lines

lilyspiniolas opened the pull request:
#### f53f6698c4254d602172dfe68a25279f77901c9d
<pre>
[iOS] Displayed values for date controls may cause overflow onto multiple lines
<a href="https://bugs.webkit.org/show_bug.cgi?id=290080">https://bugs.webkit.org/show_bug.cgi?id=290080</a>
<a href="https://rdar.apple.com/146902737">rdar://146902737</a>

Reviewed by Aditya Keerthi.

Certain wide values for date controls could cause the displayed string to overflow onto
a second line. This was caused by incorrect logic in `LocalizedDateCache::calculateMaximumWidth`
which returned the width of an arbitrary date based on the assumption numbers typically have equal
width, which is not true. The method has been updated to estimate the maximum width based on the
width of each number 0-9. It includes logic to handle localized numerals and a variety of different
date and time formats. The name of the method and its caller were update to reflect the fact that
this is just an estimation as factors such as kerning are not taken into account.

The value returned by this method is now used as the `min-width` for the date/time controls
instead of setting `width` to the value. This is done so that, even in the case where the
estimated maximum width is incorrect, the displayed value will still be on one line as the
control will now expand.

Updated the name of `LocalizedDateCache::calculatedMaximumWidthForWeek` to
`estimatedMaximumWidthForWeek` to reflect that kerning is not accounted for. Removed
an unnecessary `adoptNS()` from the method. Added logic so that it too considers the
width of localized numerals rather than only Arabic numerals.

Added a layout test which ensures that the estimated maximum width is always correct for
the system font at the default size, and that the displayed value is always displayed on
a single line.

Updated iOS test expectations for fast/forms/ios/initial-size-for-date-time-related-controls.html
since the control width has changed slightly since its width is based on estimatedMaximumWidth,
which is no longer using an arbitrary time.

* LayoutTests/fast/forms/ios/initial-size-for-date-time-related-controls-expected.txt: Added.
* LayoutTests/fast/forms/ios/initial-size-for-date-time-related-controls.html: Added.
* LayoutTests/platform/ios/fast/forms/time/time-input-rendering-basic-expected.txt:
* Source/WebCore/platform/text/cocoa/LocalizedDateCache.h:
* Source/WebCore/platform/text/cocoa/LocalizedDateCache.mm:
(WebCore::LocalizedDateCache::estimatedMaximumWidthForDateType):
(WebCore::estimateMaximumWidthForWeek):
(WebCore::LocalizedDateCache::estimateMaximumWidth):
(WebCore::LocalizedDateCache::maximumWidthForDateType): Deleted.
(WebCore::calculateMaximumWidthForWeek): Deleted.
(WebCore::LocalizedDateCache::calculateMaximumWidth): Deleted.
* Source/WebCore/rendering/ios/RenderThemeIOS.mm:
(WebCore::adjustInputElementButtonStyle):

Canonical link: <a href="https://commits.webkit.org/292441@main">https://commits.webkit.org/292441@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/37f98e9d4d5aa5f37615f7668f08fc3b34d68017

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/96048 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/15662 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/5612 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/101111 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/46557 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/98093 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/15957 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/24094 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/73222 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/30445 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/99051 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/11947 "Found 1 new test failure: fast/forms/ios/focus-input-via-button.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/86760 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/53559 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/11691 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/4512 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/45892 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/81840 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/4613 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/103139 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/23115 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/16843 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/82264 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/23366 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/82773 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/81636 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/26237 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/3676 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/16471 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/15457 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/23078 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/28233 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/22737 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/26217 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/24478 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/platform/text/cocoa/LocalizedDateCache.mm:
```diff
@@ -133,25 +133,28 @@ static void _localeChanged(CFNotificationCenterRef, void*, CFStringRef, const vo
-    for (auto [i, numLength] : indexedRange(numLengths)) {
-        numLength = measurer.measureText(makeString(i));
-        ASSERT(numLengths[i] == numLength);
+    auto allDigitNumber = adoptNS([[NSNumber alloc] initWithInteger:9876543210]);
```
```suggestion
    RetainPtr allDigitNumber = adoptNS([[NSNumber alloc] initWithInteger:9876543210]);
```

pxlcoder commented on Source/WebCore/platform/text/cocoa/LocalizedDateCache.mm:
```diff
@@ -133,25 +133,28 @@ static void _localeChanged(CFNotificationCenterRef, void*, CFStringRef, const vo
+
+    std::array<float, 10> numeralLengths;
+    for (auto [i, numLength] : indexedRange(numeralLengths)) {
+        numLength = measurer.measureText(String([localizedDigits  substringWithRange:NSMakeRange(9 - i, 1)]));
```
```suggestion
        numLength = measurer.measureText(String([localizedDigits substringWithRange:NSMakeRange(9 - i, 1)]));
```

pxlcoder commented on Source/WebCore/platform/text/cocoa/LocalizedDateCache.mm:
```diff
@@ -164,49 +167,193 @@ static float calculateMaximumWidthForWeek(const MeasureTextClient& measurer)
-        maximumWidth = std::max(maximumWidth, measurer.measureText(String(formattedDate)));
+
+    // Create our initial date components using arbitrary values. These will be updated as needed later on.
+    auto components = adoptNS([[NSDateComponents alloc] init]);
```
```suggestion
    RetainPtr components = adoptNS([[NSDateComponents alloc] init]);
```

pxlcoder commented on Source/WebCore/platform/text/cocoa/LocalizedDateCache.mm:
```diff
@@ -164,49 +167,193 @@ static float calculateMaximumWidthForWeek(const MeasureTextClient& measurer)
+    [components setYear:2025];
+
+    // Retreive localized numerals to measure.
+    auto allDigitNumber = adoptNS([[NSNumber alloc] initWithInteger:9876543210]);
```
```suggestion
    RetainPtr allDigitNumber = adoptNS([[NSNumber alloc] initWithInteger:9876543210]);
```

pxlcoder commented on Source/WebCore/platform/text/cocoa/LocalizedDateCache.mm:
```diff
@@ -164,49 +167,193 @@ static float calculateMaximumWidthForWeek(const MeasureTextClient& measurer)
+
+    // Retreive localized numerals to measure.
+    auto allDigitNumber = adoptNS([[NSNumber alloc] initWithInteger:9876543210]);
+    RetainPtr localizedDigits = [NSNumberFormatter localizedStringFromNumber:allDigitNumber.get()  numberStyle:NSNumberFormatterNoStyle];
```
```suggestion
    RetainPtr localizedDigits = [NSNumberFormatter localizedStringFromNumber:allDigitNumber.get() numberStyle:NSNumberFormatterNoStyle];
```

pxlcoder commented on LayoutTests/fast/forms/ios/initial-size-for-date-time-related-controls.html:
```diff
@@ -0,0 +1,59 @@
+
+function elementHeight(el) {
+    return el.getBoundingClientRect().bottom - el.getBoundingClientRect().y;
+}
```
Can we just use `clientHeight` and `clientWidth`?

https://developer.mozilla.org/en-US/docs/Web/API/Element/clientHeight

pxlcoder commented on Source/WebCore/platform/text/cocoa/LocalizedDateCache.mm:
```diff
@@ -164,49 +167,193 @@ static float calculateMaximumWidthForWeek(const MeasureTextClient& measurer)
+    if (shouldCalculateWidthForMonthAndYear) {
+        // Find the widest month.
+        unsigned widestMonth = 1;
+        unsigned secondWidestMonth = 2;
```
Builds are failing since this variable is never read from.

pxlcoder commented on Source/WebCore/platform/text/cocoa/LocalizedDateCache.mm:
```diff
@@ -164,49 +167,194 @@ static float calculateMaximumWidthForWeek(const MeasureTextClient& measurer)
-    // date doesn't adjust for the current timezone. This is an arbitrary date
-    // (x-27-2007) and time (10:45 PM).
-    RetainPtr<NSCalendar> gregorian = adoptNS([[NSCalendar alloc] initWithCalendarIdentifier:NSCalendarIdentifierGregorian]);
+    auto gregorian = adoptNS([[NSCalendar alloc] initWithCalendarIdentifier:NSCalendarIdentifierGregorian]);
```
```suggestion
    RetainPtr gregorian = adoptNS([[NSCalendar alloc] initWithCalendarIdentifier:NSCalendarIdentifierGregorian]);
```

pxlcoder commented on Source/WebCore/platform/text/cocoa/LocalizedDateCache.mm:
```diff
@@ -164,49 +167,204 @@ static float calculateMaximumWidthForWeek(const MeasureTextClient& measurer)
+    [components setMonth:3];
+    [components setYear:2025];
+
+    // Retreive localized numerals to measure.
```
```suggestion
    // Retrieve localized numerals to measure.
```

pxlcoder commented on Source/WebCore/platform/text/cocoa/LocalizedDateCache.mm:
```diff
@@ -164,49 +167,204 @@ static float calculateMaximumWidthForWeek(const MeasureTextClient& measurer)
+            // If the symbols used for PM are wider than AM, shift the hour forward by 12.
+            if (measurer.measureText(String([dateFormatter AMSymbol])) < measurer.measureText(String([dateFormatter PMSymbol])))
+                hourCandidate += 12;
+
```
```suggestion
```

pxlcoder commented on Source/WebCore/platform/text/cocoa/LocalizedDateCache.mm:
```diff
@@ -164,49 +167,204 @@ static float calculateMaximumWidthForWeek(const MeasureTextClient& measurer)
+            float hourCandidateLength = numeralLengths[widestNumZeroThroughOne] + numeralLengths[widestNum];
+            if (hourCandidateLength < numeralLengths[2] + numeralLengths[widestNumZeroThroughThree])
+                hourCandidate = 20 + widestNumZeroThroughThree;
+
```
```suggestion
```

lilyspiniolas committed:
[iOS] Displayed values for date controls may cause overflow onto multiple lines
https://bugs.webkit.org/show_bug.cgi?id=290080
rdar://146902737

Reviewed by Aditya Keerthi.

Certain wide values for date controls could cause the displayed string to overflow onto
a second line. This was caused by incorrect logic in `LocalizedDateCache::calculateMaximumWidth`
which returned the width of an arbitrary date based on the assumption numbers typically have equal
width, which is not true. The method has been updated to estimate the maximum width based on the
width of each number 0-9. It includes logic to handle localized numerals and a variety of different
date and time formats. The name of the method and its caller were update to reflect the fact that
this is just an estimation as factors such as kerning are not taken into account.

The value returned by this method is now used as the `min-width` for the date/time controls
instead of setting `width` to the value. This is done so that, even in the case where the
estimated maximum width is incorrect, the displayed value will still be on one line as the
control will now expand.

Updated the name of `LocalizedDateCache::calculatedMaximumWidthForWeek` to
`estimatedMaximumWidthForWeek` to reflect that kerning is not accounted for. Removed
an unnecessary `adoptNS()` from the method. Added logic so that it too considers the
width of localized numerals rather than only Arabic numerals.

Added a layout test which ensures that the estimated maximum width is always correct for
the system font at the default size, and that the displayed value is always displayed on
a single line.

Updated iOS test expectations for fast/forms/ios/initial-size-for-date-time-related-controls.html
since the control width has changed slightly since its width is based on estimatedMaximumWidth,
which is no longer using an arbitrary time.

* LayoutTests/fast/forms/ios/initial-size-for-date-time-related-controls-expected.txt: Added.
* LayoutTests/fast/forms/ios/initial-size-for-date-time-related-controls.html: Added.
* LayoutTests/platform/ios/fast/forms/time/time-input-rendering-basic-expected.txt:
* Source/WebCore/platform/text/cocoa/LocalizedDateCache.h:
* Source/WebCore/platform/text/cocoa/LocalizedDateCache.mm:
(WebCore::LocalizedDateCache::estimatedMaximumWidthForDateType):
(WebCore::estimateMaximumWidthForWeek):
(WebCore::LocalizedDateCache::estimateMaximumWidth):
(WebCore::LocalizedDateCache::maximumWidthForDateType): Deleted.
(WebCore::calculateMaximumWidthForWeek): Deleted.
(WebCore::LocalizedDateCache::calculateMaximumWidth): Deleted.
* Source/WebCore/rendering/ios/RenderThemeIOS.mm:
(WebCore::adjustInputElementButtonStyle):

Canonical link: https://commits.webkit.org/292441@main
